include/kodlab_mjbots_sdk/pid.h

```diff
@@ -0,0 +1,192 @@
+/**
+ * @file pid.h
+ * @author Chandravaran Kunjeti (kunjeti@seas.upenn.edu)
+ * @brief Class implementation of PID
+ * @date 9/30/2022
+ *
+ * @copyright 2022 The Trustees of the University of Pennsylvania. All rights reserved.
+ *
+ */
+
+#pragma once
+
+#ifndef _PID_
+#define _PID_
+
+#include "Eigen/Geometry"
+
+namespace kodlab
+{
+
+  /**
+   * @brief Pid object
+   */
+  class Pid
+  {
+
+  public:
+    /**
+     * @brief Construct a Pid object
+     * @param p_gain Proportional Gain
+     * @param d_gain Derivative Gain
+     * @param i_gain Integral Gain
+     * @param time_step Time difference between 2 updates
+     */
+    Pid(double p_gain, double d_gain, double i_gain, double time_step, double deadband, double saturation)
+        : kp_(0), kd_(0), ki_(0), time_step_(0.01), deadband_(0), saturation_(0)
+    {
+
+      kp_ = p_gain;
+      kd_ = d_gain;
+      ki_ = i_gain;
+      time_step_ = time_step;
+      deadband_ = deadband;
+      saturation_ = saturation;
+    }
+
+    /**
+     * @brief Function to set the PID gains
+     * @param p_gain Proportional Gain
+     * @param d_gain Derivative Gain
+     * @param i_gain Integral Gain
+     */
+    void SetGains(double p_gain, double d_gain, double i_gain)
+    {
+
+      kp_ = p_gain;
+      kd_ = d_gain;
+      ki_ = i_gain;
+    }
+
+    /**
+     * @brief Function to set the deadband
+     * @param deadband Deadband that needs to be set
+     */
+    void SetDeadband(double deadband)
+    {
+
+      deadband_ = deadband;
+    }
+
+    /**
+     * @brief Function to set the Saturation
+     * @param saturation Saturation
+     */
+    void SetSaturation(double saturation)
+    {
+
+      saturation_ = saturation;
+    }
+
+    /**
+     * @brief Function to update the PID output.
+     * @param goal Final goal state for the controller.
+     * @param current Current state of the controller.
+     * @return Returns the Pid output
+     */
+    double Update(double goal, double current)
+    {
+
+      // Calculate the errors
+      // Proportional error
+      error_ = goal - current;
+
+      // Integral error
+      i_error_ += error_ * time_step_;
+
+      // Derivative error
+      derivative_ = (error_ - d_error_) / time_step_;
+
+      // Control output
+      output_ = kp_ * error_ + kd_ * derivative_ + ki_ * i_error_;
+
+      // Saturation checking
+      output_ = (output_ > saturation_) ? saturation_ : output_;
```

Comment:
This only handles positive saturation.

Reply:
Should I make saturation an array, then we could set the lower and upper limit of the saturation?

Reply:
It is unclear what you are asking here... 

I am simply saying that the current implementation only saturates positive output.  Any saturation functionality should cover both positive and negative outputs.

As written, with a saturation torque of `10 N-m`, 
| Ctrl. Torque | Output Torque |
|--------------|---------------|
|   5          |   5           |
|  15          |  10           |
| -15          | -15           |

Reply:
Okay I get what you are saying. So I was thinking in a slightly different direction. So let us say we have a limit for a positive saturation as 10. This doesn't necessarily mean my lower bound would be -10. It could be some different value like 5. So I was thinking of making saturation an array that allows you to give 2 values, a lower bound and an upper bound. I have the implementation based on how I have explained it above. If you need me to change it to single saturation value let me know.

Reply:
I agree that actuator saturation is not necessarily symmetric, and accepting upper and lower bounds makes sense.  But the implementation at the time of writing only clipped the upper end.

Reply:
Correct, I have implemented the array based implementation.